lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -402,9 +403,228 @@ struct ConvolutionToConv2dPattern
   }
 };
 
+struct PoolingToPool2dPattern : public OpConversionPattern<ttir::PoolingOp> {
+public:
+  using OpConversionPattern<ttir::PoolingOp>::OpConversionPattern;
+
+  std::vector<int64_t> getSpatialDims(ttir::PoolingOp op) const {
+    std::vector<int64_t> spatial_dims;
+    for (int64_t i = 0;
+         i < static_cast<int64_t>(op.getWindowDimensions().size()); i++) {
+      if (op.getWindowDimensions()[i] > 1) {
+        spatial_dims.push_back(i);
+      }
+    }
+    return spatial_dims;
+  }
+
+  LogicalResult is2d(ttir::PoolingOp op) const {
+
+    // Window dimensions must be 4 in length
+    if (op.getWindowDimensions().size() != 4) {
+      return failure();
+    }
+
+    // Window strides must be 4 in length
+    if (op.getWindowStrides().size() != 4) {
+      return failure();
+    }
+
+    // Operand rank(s) must be 4
+    for (Value operand : op.getInputs()) {
+      auto operand_type = mlir::cast<mlir::RankedTensorType>(operand.getType());
+      if (operand_type.getRank() != 4) {
+        return failure();
+      }
+    }
+
+    // Exactly two of the window dimensions must be greater than 1
+    std::vector<int64_t> true_window_dimensions_indices = getSpatialDims(op);
+
+    if (true_window_dimensions_indices.size() != 2) {
+      return failure();
+    }
+
+    // Exactly two of the window strides must be greater than 1
+    std::vector<int64_t> true_window_stride_indices;
+    for (int64_t i = 0; i < static_cast<int64_t>(op.getWindowStrides().size());
+         i++) {
+      if (op.getWindowStrides()[i] > 1) {
+        true_window_stride_indices.push_back(i);
+      }
+    }
+
+    if (true_window_stride_indices.size() != 2) {
+      return failure();
+    }
+
+    // The indices of the true window dimensions and strides must be the same
+    if ((true_window_dimensions_indices[0] != true_window_stride_indices[0] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[1]) &&
+        (true_window_dimensions_indices[0] != true_window_stride_indices[1] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[0])) {
+      return failure();
+    }
+
+    // NOTE: To model later
```

Comment:
We should have an issue tracking todos

Reply:
I think Im going to go more with the rewrite failure you mentioned above. I think we want to go with supporting maxpool methods as we go. As of right now I believe ttnn only supports maxpool with square windows, which is a much easier thing to look for.